Self-Consistent-Field/RHF_symmetry.py

```diff
@@ -0,0 +1,192 @@
+
+# Memory for NumPy in GB
+numpy_memory = 2
+
+mol = psi4.geometry("""
+O
+H 1 1.1
+H 1 1.1 2 104
+""")
+
+psi4.set_options({'basis': 'sto-3g',
+                  'scf_type': 'direct',
+                  'e_convergence': 1e-8})
+
+# Set defaults
+maxiter = 50
+E_conv = 1.0E-8
+D_conv = 1.0E-7
+
+# Integral generation from Psi4's MintsHelper, which automatically
+# performs symmetry adaptation.
+wfn = psi4.core.Wavefunction.build(mol, psi4.core.get_global_option('BASIS'))
+t = time.time()
+mints = psi4.core.MintsHelper(wfn.basisset())
+
+nirrep = wfn.nirrep()
+# Water belongs to the C2v point group, which has four irreducable
+# representations (irreps): A_1, A_2, B_1, and B_2.
+assert nirrep == 4
+nsopi = list(wfn.nsopi())
+
+# Get nbf and ndocc for closed shell molecules
+nbf = sum(nsopi)
+ndocc = wfn.nalpha()
+
+print('\nNumber of occupied orbitals: %d' % ndocc)
+print('Number of basis functions: %d' % nbf)
+print('Number of spin orbitals per irrep:', nsopi)
+
+# Run a quick check to make sure everything will fit into memory
+I_Size = (nbf**4) * 8.e-9
+print("\nSize of the ERI tensor will be %4.2f GB." % I_Size)
+
+# Estimate memory usage
+memory_footprint = I_Size * 1.5
+if I_Size > numpy_memory:
+    psi4.core.clean()
+    raise Exception("Estimated memory utilization (%4.2f GB) exceeds numpy_memory "
+                    "limit of %4.2f GB." % (memory_footprint, numpy_memory))
+
+# A matrix is returned for every irrep, even if the irrep is not present for
+# the molecule; these must be filtered out to avoid problems with NumPy arrays
+# that have a zero dimension.
+def filter_empty_irrep(coll):
+    return tuple(m for m in coll if all(m.shape))
+
+# The convention will be to have quantities in the spin orbital (SO) basis
+# ending with an underscore, each consisting of one matrix per irrep.
+S_ = filter_empty_irrep(mints.so_overlap().to_array())
+T_ = filter_empty_irrep(mints.so_kinetic().to_array())
+V_ = filter_empty_irrep(mints.so_potential().to_array())
+# The two-electron integrals are not blocked according to symmetry, so a
+# transformation between the atomic orbital (AO) and SO bases will be required.
+I_AO = np.asarray(mints.ao_eri())
+# In order to convert from the C1 AO basis to the symmetrized SO basis, a set
+# of matrices (one for each irrep with shape [nao, irrep_size]) is used to
+# transform the dense AO representation into the block-diagonal SO
+# representation. A SO-to-AO transformation matrix is obtained by transposing
+# the corresponding AO-to-SO transformation matrix.
+transformers_ = filter_empty_irrep(wfn.aotoso().to_array())
+
+assert len(S_) == len(T_) == len(V_) == len(transformers_)
+nirrep = len(S_)
+nsopi = [n for n in nsopi if n > 0]
+
+print('\nTotal time taken for integrals: %.3f seconds.' % (time.time() - t))
+t = time.time()
+
+# Build H_core: [Szabo:1996] Eqn. 3.153, pp. 141
+H_ = [T + V for (T, V) in zip(T_, V_)]
+
+# Build the orthogonalizer A = S^(-1/2) for each irrep.
```

Comment:
This might be a good case to use the `Matrix.power` Psi4 function which both handles the irreps and protects small eigenvalues in the power function. Not sure if that makes the code more clear or less, your call there I think.

Reply:
I figured that because `Matrix.power` is used in `RHF.py`, this could be an example of how it's done with pure NumPy.